tensorflow/contrib/verbs/rdma_mgr.cc

```diff
@@ -115,6 +115,99 @@ void RdmaMgr::SetupChannels() {
   }
 }
 
+#define PING_RECV_WRID 0
+#define PING_BUFF_SIZE 1024
+
+int RdmaMgr::PostRecv(RdmaChannel* rc, struct ibv_sge list) {
+  struct ibv_recv_wr wr, *bad_wr;
+  memset(&wr, 0, sizeof(wr));
+  wr.sg_list = &list;
+  wr.num_sge = 1;
+  wr.wr_id = PING_RECV_WRID;
+
+  return ibv_post_recv(rc->qp_, &wr, &bad_wr);
+}
+
+int RdmaMgr::PostSend(RdmaChannel* rc, struct ibv_sge list) {
+  struct ibv_send_wr wr, *bad_wr;
+  memset(&wr, 0, sizeof(wr));
+  wr.wr_id = (uint64_t)rc;
+  wr.sg_list = &list;
+  wr.num_sge = 1;
+  wr.opcode = IBV_WR_SEND;
+  wr.send_flags = IBV_SEND_SIGNALED;
+
+  return ibv_post_send(rc->qp_, &wr, &bad_wr);
+}
+
+// Check connectivity by pinging every channel
+bool RdmaMgr::ConnectivityCheck() {
+  int i, rcnt = 0, scnt = 0;
+  void* buff;
+  struct ibv_sge list;
+  buff = malloc(PING_BUFF_SIZE);
+  CHECK(buff) << "Malloc failed!";
+  struct ibv_mr* mr = ibv_reg_mr(rdma_adapter_->pd_, buff, PING_BUFF_SIZE,
+                                 IBV_ACCESS_LOCAL_WRITE);
+  CHECK(mr) << "Failed to register memory region";
+
+  memset(&list, 0, sizeof(list));
+  list.addr = (uintptr_t)buff;
+  list.length = PING_BUFF_SIZE;
+  list.lkey = mr->lkey;
+
+  for (const auto& p : channel_table_) {
+    string worker_name = p.first;
+    RdmaChannel* rc = p.second;
+
+    VLOG(2) << "Ping to " << worker_name;
+    CHECK(PostRecv(rc, list) == 0) << "Couldn't post receive from "
```

Comment:
Just one thing, I think we need to deal with the scenario that One server is delayed between the ending of SetupChannels and until it comes in here. then we do post send to someone that haven't done post recv and the packet will be lost. one possible way is to do post receive before the channels connections (so it will act like a barrier) and then do only post send in this function. what do u think ?